### PR TITLE
resolve the snapshot boom problem caused by cdc incremental scans

### LINKED ISSUE #15910: Number of snapshots  increases tremendously and TiKV OOM when resuming changefeed with a ts 6 days before.
### What version of TiKV are you using?
Release Version:   6.5.5                                       
Edition:           Community                                   
Git Commit Hash:   7f026c027741ef98204959c37e10a7de62934480    

### What operating system and CPU are you using?
K8S
/ # cat /etc/*-release                           
3.14.6                                           
NAME="Alpine Linux"                              
ID=alpine                                        
VERSION_ID=3.14.6                                
PRETTY_NAME="Alpine Linux v3.14"                 
HOME_URL="https://alpinelinux.org/"              
BUG_REPORT_URL="https://bugs.alpinelinux.org/"   

### Steps to reproduce
- TiDB GC is stopped for several days. (MVCC about 2TB, from 2.6T to 4.6T)
- There is a cdc changefeed running
- TiKV configuration cdc.incremental-fetch-speed-limit set to 128MiB
- Resume CDC changefeed and overwirte checkpoint ts with a ts 6 days before

### What did you expect?
TiKV not OOM

### What did happened?
Numbers of Snapshot increase tremendously to million level, and TiKV OOM is seen.
After pause cdc changefeed, snapshot number decreased as well
image
image

### PR DESCRIPTION
**Pull request by hicqu**
### What is changed and how it works?

Issue Number: Close #15910 

The problem is introduced in #15865 .

What's Changed:

```commit-message
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/cdc/src/initializer.rs [hunk 1]
```diff
@@ -207,12 +204,24 @@ impl<E: KvEngine> Initializer<E> {
...
-        }
+        let scan_concurrency_semaphore = self.scan_concurrency_semaphore.clone();
+        let mut downstream_closed = Box::new(future::pending().shared())
+            as Box<dyn Future<Output = ()> + Send + Sync + Unpin>;
+        swap(&mut self.downstream_closed, &mut downstream_closed);
+        let _permit = select! {
+            _ = (&mut downstream_closed).fuse() => {
+                return on_cancel();
+            }
+            x = scan_concurrency_semaphore.acquire().fuse() => match x {
```
Comment (3pointer) [inline components/cdc/src/initializer.rs:215]: if block on `scan_concurrency_semaphore` here, will CDC client detect the block and stop subscribe outside?

Comment (hicqu) [reply components/cdc/src/initializer.rs:215]: Yes, the branch is in a `select` block.

**Review by overvenus — COMMENTED** [review]
Why not revert #15865? It essentially undo #10262 which not only solves too many snapshot issue but also solve an OOM issue caused by too many concurrent tasks.

**Comment by hicqu** [comment]
> Why not revert #15865? It essentially undo #10262 which not only solves too many snapshot issue but also solve an OOM issue caused by too many concurrent tasks.

@overvenus Without #15865, lots of requests can be blocked on `scan_concurrent_semaphore`. So if those regions get changed, TiCDC clients can't be notified quickly.

**Comment by pingyu** [comment]
Is it still possible to get too many snapshots and OOM ?

Additionally, I have reviewed issue #15866 and I'm curious if it's feasible to still acquire the semaphore at the beginning of the `initialize()` process, but also receive notifications and cancel the initialization task when the region changes.

**Pull request closed without merging**